File: src/0024_swap_nodes_in_pairs.js
function ListNode(val = 0, next = null) {
    this.val = val;
    this.next = next;
}

function swapPairs(head) {
    const dummy = new ListNode(0, head);
    let prev = dummy;
    let curr = head;
    while (curr && curr.next) {
        const nxtPair = curr.next.next;
        const second = curr.next;
        second.next = curr;
        curr.next = nxtPair;
        prev.next = second;
        prev = curr;
        curr = nxtPair;
    }
    return dummy.next;
}

module.exports = { swapPairs };

if (require.main === module) {
    const assert = require('assert');
    function ListNode(val = 0, next = null) { this.val = val; this.next = next; }
    function TreeNode(val = 0, left = null, right = null) { this.val = val; this.left = left; this.right = right; }
    function Node(val = 0, neighbors = []) { this.val = val; this.neighbors = neighbors; }
    function listNodeToString(node) {
        const values = [];
        let current = node;
        while (current !== null) {
            values.push(String(current.val));
            current = current.next;
        }
        return values.join('->');
    }
    function treeToString(node) {
        if (node === null) return 'None';
        return `${node.val}(${treeToString(node.left)},${treeToString(node.right)})`;
    }
    function sorted(value) { return value.slice().sort(); }
    assert.deepStrictEqual(listNodeToString(swapPairs(new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, null)))))), listNodeToString(new ListNode(2, new ListNode(1, new ListNode(4, new ListNode(3, null))))));
    assert.deepStrictEqual(swapPairs(null), null);
    assert.deepStrictEqual(listNodeToString(swapPairs(new ListNode(1, null))), listNodeToString(new ListNode(1, null)));
}

File: src/0024_swap_nodes_in_pairs.rs
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { val, next: None }
    }
}

struct Solution;

impl Solution {
    pub fn swap_pairs(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        match head {
            Some(mut first) => match first.next.take() {
                Some(mut second) => {
                    let next_pair = second.next.take();
                    first.next = Self::swap_pairs(next_pair);
                    second.next = Some(first);
                    Some(second)
                }
                None => Some(first),
            },
            None => None,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn list(values: &[i32]) -> Option<Box<ListNode>> {
        values
            .iter()
            .rev()
            .fold(None, |next, &val| Some(Box::new(ListNode { val, next })))
    }

    fn list_values(head: &Option<Box<ListNode>>) -> Vec<i32> {
        let mut values = Vec::new();
        let mut current = head.as_deref();
        while let Some(node) = current {
            values.push(node.val);
            current = node.next.as_deref();
        }
        values
    }

    #[test]
    fn mirrors_python_main_assertions() {
        assert_eq!(
            list_values(&Solution::swap_pairs(list(&[1, 2, 3, 4]))),
            vec![2, 1, 4, 3]
        );
        assert!(Solution::swap_pairs(None).is_none());
        assert_eq!(list_values(&Solution::swap_pairs(list(&[1]))), vec![1]);
    }
}
